amads/melody/contour/interpolation_contour.py

```diff
@@ -0,0 +1,314 @@
+    (2004) [3].
+    """
+
+    def __init__(self, times: list[float], pitches: list[int]):
+        """Initialize with time and pitch values.
+
+        Parameters
+        ----------
+        times : list[float]
+            Array of onset times in seconds
+        pitches : list[int]
+            Array of pitch values
+
+        Raises
+        ------
+        ValueError
+            If times and pitches are not the same length
+
+        Examples
+        --------
+        >>> happy_birthday_pitches = [
+        ...     60, 60, 62, 60, 65, 64, 60, 60, 62, 60, 67, 65,
+        ...     60, 60, 72, 69, 67, 65, 64, 70, 69, 65, 67, 65
+        ... ]
+        >>> happy_birthday_times = [
+        ...     0, 0.75, 1, 2, 3, 4, 6, 6.75, 7, 8, 9, 10,
+        ...     12, 12.75, 13, 14, 15, 16, 17, 18, 18.75, 19, 20, 21
+        ... ]
+        >>> ic = InterpolationContour(happy_birthday_times, happy_birthday_pitches)
+        >>> ic.direction_changes
+        0.6
+        >>> ic.class_label
+        'ccbc'
+        >>> ic.mean_gradient
+        2.512...
+        >>> ic.gradient_std
+        5.496...
+        >>> ic.global_direction
+        1
+
+        References
+        ----------
+        [1] Müllensiefen, D. (2009). Fantastic: Feature ANalysis Technology Accessing
+        STatistics (In a Corpus): Technical Report v1.5
+        [2] W. Steinbeck, Struktur und Ähnlichkeit: Methoden automatisierter
+            Melodieanalyse. Bärenreiter, 1982.
+        [3] Müllensiefen, D. & Frieler, K. (2004). Cognitive Adequacy in the Measurement
+        of Melodic Similarity: Algorithmic vs. Human Judgments
+        """
+        if len(times) != len(pitches):
+            raise ValueError("Times and pitches must have the same length")
+        self.times = times
+        self.pitches = pitches
+        self.contour = self.calculate_interpolation_contour()
+
+    @staticmethod
+    def _find_contour_extrema(pitches: list[int]) -> list[int]:
+        """Determine contour extremum notes (local minima and maxima, including endpoints).
+
+        Excludes changing notes (notae cambiatae).
+
+        Parameters
+        ----------
+        pitches : list[int]
+            List of MIDI pitch values
+
+        Returns
+        -------
+        list[int]
+            Array of indices corresponding to extrema points in the melody
+        """
+        extrema_indices = []
+        n = len(pitches)
+
+        # Always include first and last notes
+        extrema_indices.append(0)
+
+        # Find local extrema
+        for i in range(1, n - 1):
+            # Check previous and next notes
+            prev_pitch = pitches[i - 1]
+            curr_pitch = pitches[i]
+            next_pitch = pitches[i + 1]
+
+            # Skip changing notes (where adjacent pitches are equal)
+            if prev_pitch == next_pitch:
+                continue
+
+            # Check if it's a peak or valley
+            is_peak = prev_pitch < curr_pitch and next_pitch < curr_pitch
```

Comment:
This peak detection logic would best be done by first removing duplicate adjacent notes, and then running your code

```py
            is_peak = prev_pitch < curr_pitch and next_pitch < curr_pitch
            is_valley = prev_pitch > curr_pitch and next_pitch > curr_pitch
```

and then skipping the "Additional check for equal adjacent notes".

Proposal we agreed on: support both behaviours, but choose which one depending on the `method` parameter.

Reply:
implemented